core/compiler.common.jvm/src/org/jetbrains/kotlin/load/java/AbstractAnnotationTypeQualifierResolver.kt

```diff
@@ -0,0 +1,205 @@
+        return knownNullability(typeQualifierAnnotation, typeQualifierAnnotation.forceWarning())
+            ?.copy(isForWarningOnly = jsr305State.isWarning)
+    }
+
+    fun extractNullability(
+        annotations: Iterable<Annotation>, forceWarning: Annotation.() -> Boolean = { false }
+    ): NullabilityQualifierWithMigrationStatus? =
+        annotations.firstNotNullOfOrNull { extractNullability(it, forceWarning) }
+
+    fun extractMutability(annotations: Iterable<Annotation>): MutabilityQualifier? {
+        return annotations.fold(null as MutabilityQualifier?) { found, annotation ->
+            when (annotation.fqName) {
+                in READ_ONLY_ANNOTATIONS -> MutabilityQualifier.READ_ONLY
+                in MUTABLE_ANNOTATIONS -> MutabilityQualifier.MUTABLE
+                else -> found
+            }.also { if (found != null && found != it) return null /* inconsistent */ }
+        }
+    }
+
+    private fun extractDefaultQualifiers(annotation: Annotation): JavaDefaultQualifiers? {
+        resolveQualifierBuiltInDefaultAnnotation(annotation)?.let { return it }
+
+        val (typeQualifier, applicability) = resolveTypeQualifierDefaultAnnotation(annotation)
+            ?: return null
+        val jsr305State = resolveJsr305CustomState(annotation) ?: resolveJsr305AnnotationState(typeQualifier)
+        if (jsr305State.isIgnore) return null
+        // TODO: since we override the warning status, whether we force it in `extractNullability` is irrelevant.
+        //   However, this is probably not what was intended.
+        val nullabilityQualifier = extractNullability(typeQualifier) { false } ?: return null
+        return JavaDefaultQualifiers(nullabilityQualifier.copy(isForWarningOnly = jsr305State.isWarning), applicability)
+    }
+
+    fun extractAndMergeDefaultQualifiers(
+        oldQualifiers: JavaTypeQualifiersByElementType?, annotations: Iterable<Annotation>
+    ): JavaTypeQualifiersByElementType? {
+        if (javaTypeEnhancementState.disabledDefaultAnnotations) return oldQualifiers
+
+        val defaultQualifiers = annotations.mapNotNull { extractDefaultQualifiers(it) }
+        if (defaultQualifiers.isEmpty()) return oldQualifiers
+
+        val defaultQualifiersByType =
+            oldQualifiers?.defaultQualifiers?.let(::QualifierByApplicabilityType)
+                ?: QualifierByApplicabilityType(AnnotationQualifierApplicabilityType::class.java)
+
+        var wasUpdate = false
+        for (qualifier in defaultQualifiers) {
+            for (applicabilityType in qualifier.qualifierApplicabilityTypes) {
+                defaultQualifiersByType[applicabilityType] = qualifier
+                wasUpdate = true
+            }
+        }
+
+        return if (!wasUpdate) oldQualifiers else JavaTypeQualifiersByElementType(defaultQualifiersByType)
+    }
+
+    private fun knownNullability(annotation: Annotation, forceWarning: Boolean): NullabilityQualifierWithMigrationStatus? {
```

Comment:
BTW Why we need `forceWarning` and can't use the corresponding report level for the specific annotation instead?

Reply:
Because certain annotations only cause warnings regardless of what the annotation itself says, e.g. with TypeEnhancementImprovementsInStrictMode disabled. Like I said above, I don't yet know how to implement this in FIR.

Reply:
Got it. Thanks. It seems we may not to support disabled `TypeEnhancementImprovementsInStrictMode` by the time FIR is released. So I'd suggest to consider `TypeEnhancementImprovementsInStrictMode` as permanently enabled for FIR (just in case cc @mglukhikh). Are there any other cases for which `forceWarning` is needed?

Reply:
>Are there any other cases for which forceWarning is needed?

Yes, there's also this branch: https://github.com/JetBrains/kotlin/blob/8f484fc/core/descriptors.jvm/src/org/jetbrains/kotlin/load/java/typeEnhancement/signatureEnhancement.kt#L96-L101

I don't know that this means.